File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.ts
function successfulPairs(spells: number[], potions: number[], success: number): number[] {
    const result = [];
    potions.sort((a,b)=>a-b)

    function findFirstPair(n: number): number {
        let start = 0;
        let end = potions.length - 1;

        while(start < end) {
            const mid = start + Math.ceil((end - start) / 2)        
            if(potions[mid] * n >= success) {
                end = mid - 1;
            } else {
                start = mid;
            }
        }
        return potions[end] * n >= success ? end - 1 : end;
    }

    for(const spell of spells) {
        const first = findFirstPair(spell)
        result.push(potions.length - first - 1)
    }

    return result;
};

File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.rs
impl Solution {
    pub fn successful_pairs(spells: Vec<i32>, mut potions: Vec<i32>, success: i64) -> Vec<i32> {
        let mut res = vec![0; spells.len()];
        let n = potions.len();
        potions.sort();
        for (i, val) in spells.iter().enumerate() {
            if n == 0 { continue }
            let mut left = 0;
            let mut right = n;
            while left < right {
                let mid = (left + right) / 2;
                let mid_value = potions[mid] as i64 * *val as i64;
                if mid_value < success {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }
            res[i] = n as i32  - right as i32;
        }
        res
    }
}